### ISSUE TITLE
Keeping track of dxds in DEFunc breaks copy.deepcopy

### ISSUE THREAD

**Issue by lucaslie** [DiffEqML/torchdyn#61]
**Describe the bug**

I attempted to deepcopy a `DEFunc` but deepcopy failed because only leaf tensors can be deepcopied (see screenshot below). 

I tracked it back to the `DEFunc.dxds` attribute which seems to be just a slice of the input `x` (or the entire input) to `DEFunc.forward`, see [this line](https://github.com/DiffEqML/torchdyn/blob/43d22ce5e12dc1a298ef8c11e64f015b05a60765/torchdyn/models/defunc.py#L52).

**Step to Reproduce**

```python
import torchdyn
import torchdyn.models
import copy
import torch
import torch.nn as nn

# create Vanilla NeuralDE
f_forward = nn.Sequential(nn.Linear(2, 16), nn.Tanh(), nn.Linear(16, 2))
model = torchdyn.models.NeuralDE(
    f_forward,
    sensitivity="adjoint",
    solver="dopri5",
)

# create some fake data of batch size 1
x = torch.rand(1, 2)

# do a deepcopy, nothing fails because we never tracked dxds (no forward pass)
model_copy1 = copy.deepcopy(model)

# do a forward+backward pass
y = model(x)
loss = y.sum()
loss.backward()

# do another deepcopy, fails because model.defunc tracks dxds
model_copy2 = copy.deepcopy(model)
````

**Expected behavior**

Keeping track of pytorch tensors as plain attributes of classes can induce all kinds of unexpected behaviors. `copy.deepcopy` is just one example of that but another example would for example be if we moved the `model` from one device to another. While all properly registered tensors would also be moved to a different device `model.defunc.dxds` would be left at the original device. 

I would suggest one of the following: 

* I would properly register `dxds` as buffer using [`torch.nn.Module.register_buffer`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer) and then use a `detach().clone()` operation to track `dxds` during the forward pass without creating any memory leaks or potentially breaking any kind of upstream tasks like `copy.deepcopy`.

* I could not find any place in the codebase where `dxds` was used. So maybe just removing it all together might be easier? 

**Screenshots**

![image](https://user-images.githubusercontent.com/11156568/103749175-2ad18500-4fd3-11eb-8c74-714f9731129b.png)

**Comment by Zymrael** [comment]
Thank you for spotting the problem and for the extremely detailed post. I've pushed a quick fix (and a regression test to avoid similar problems in the future). As you said, the issue was caused by `self.dxds` in `DEFunc` forwards (removed for now).

 The idea at the time of implementation was to offer a quick access point to certain quantities for regularization purposes. Unfortunately, we never got around to implementing a proper API. The `DEFunc` class should be partially redesigned in the future -- this fix will do for now.

**Comment by lucaslie** [comment]
Awesome, thank you so much for the quick support. :) 

I was worried I would have to write a wrapper or fork my own version of it but this way of course it's much easier and convenient.  ;) 

Feel free to close the issue if you want!

Thanks again and Cheers, 
Lucas

**Comment by Zymrael** [comment]
No problem! Feel free to open PRs in the future, even for smaller changes. In general, we should be able to find a way to customize the core `torchdyn` without additional wrappers or user forks. 

We appreciate all kinds of contributions, and `torchdyn` certainly has many places where it can be improved.
